File: resolve_silence_cut/silence.py
from __future__ import annotations

import math
import struct
import wave
from pathlib import Path

from .config import SilenceSettings
from .models import TimeRangeSeconds


def read_wav_pcm(path: str) -> tuple[int, int, bytes]:
    wav_path = Path(path)
    with wave.open(str(wav_path), "rb") as wav_file:
        sample_rate = wav_file.getframerate()
        sample_width = wav_file.getsampwidth()
        channels = wav_file.getnchannels()
        frames = wav_file.getnframes()
        raw = wav_file.readframes(frames)

    if channels > 1:
        raw = to_mono(raw, sample_width, channels)

    return sample_rate, sample_width, raw


def rms_to_db(rms_value: float) -> float:
    if rms_value <= 0:
        return -120.0
    max_16_bit = 32768.0
    normalized = min(1.0, rms_value / max_16_bit)
    return 20.0 * math.log10(max(normalized, 1e-9))


def detect_silence_ranges(
    sample_rate: int,
    sample_width: int,
    raw_pcm: bytes,
    settings: SilenceSettings,
) -> list[TimeRangeSeconds]:
    frame_samples = max(1, int(sample_rate * settings.analysis_frame_ms / 1000.0))
    frame_bytes = frame_samples * sample_width
    min_silence_seconds = settings.min_silence_duration_ms / 1000.0

    frame_loudness_db: list[float] = []
    for idx in range(0, len(raw_pcm), frame_bytes):
        chunk = raw_pcm[idx : idx + frame_bytes]
        if not chunk:
            continue
        rms_val = rms_chunk(chunk, sample_width)
        frame_loudness_db.append(rms_to_db(float(rms_val)))

    silent_ranges: list[TimeRangeSeconds] = []
    current_start: int | None = None
    for frame_index, loudness_db in enumerate(frame_loudness_db):
        is_silent = loudness_db <= settings.silence_threshold_db
        if is_silent and current_start is None:
            current_start = frame_index
        if not is_silent and current_start is not None:
            start_sec = current_start * settings.analysis_frame_ms / 1000.0
            end_sec = frame_index * settings.analysis_frame_ms / 1000.0
            if (end_sec - start_sec) >= min_silence_seconds:
                silent_ranges.append(TimeRangeSeconds(start=start_sec, end=end_sec))
            current_start = None

    if current_start is not None:
        start_sec = current_start * settings.analysis_frame_ms / 1000.0
        end_sec = len(frame_loudness_db) * settings.analysis_frame_ms / 1000.0
        if (end_sec - start_sec) >= min_silence_seconds:
            silent_ranges.append(TimeRangeSeconds(start=start_sec, end=end_sec))

    return merge_ranges(silent_ranges)


def merge_ranges(ranges: list[TimeRangeSeconds]) -> list[TimeRangeSeconds]:
    if not ranges:
        return []
    sorted_ranges = sorted(ranges, key=lambda r: r.start)
    merged: list[TimeRangeSeconds] = [sorted_ranges[0]]
    for rng in sorted_ranges[1:]:
        prev = merged[-1]
        if rng.start <= prev.end:
            merged[-1] = TimeRangeSeconds(start=prev.start, end=max(prev.end, rng.end))
        else:
            merged.append(rng)
    return merged


def expand_with_padding(
    ranges: list[TimeRangeSeconds],
    clip_duration_seconds: float,
    padding_before_ms: int,
    padding_after_ms: int,
) -> list[TimeRangeSeconds]:
    before = max(0.0, padding_before_ms / 1000.0)
    after = max(0.0, padding_after_ms / 1000.0)
    expanded = [
        TimeRangeSeconds(
            start=max(0.0, rng.start - before),
            end=min(clip_duration_seconds, rng.end + after),
        )
        for rng in ranges
    ]
    return merge_ranges(expanded)


def invert_ranges(
    remove_ranges: list[TimeRangeSeconds],
    total_duration_seconds: float,
) -> list[TimeRangeSeconds]:
    if total_duration_seconds <= 0:
        return []
    if not remove_ranges:
        return [TimeRangeSeconds(0.0, total_duration_seconds)]

    keeps: list[TimeRangeSeconds] = []
    cursor = 0.0
    for rng in remove_ranges:
        if rng.start > cursor:
            keeps.append(TimeRangeSeconds(cursor, min(rng.start, total_duration_seconds)))
        cursor = max(cursor, rng.end)

    if cursor < total_duration_seconds:
        keeps.append(TimeRangeSeconds(cursor, total_duration_seconds))

    return [k for k in keeps if k.duration > 0.0]


def to_mono(raw_pcm: bytes, sample_width: int, channels: int) -> bytes:
    if channels <= 1:
        return raw_pcm
    samples = decode_samples(raw_pcm, sample_width)
    mono_samples: list[int] = []
    for idx in range(0, len(samples), channels):
        channel_slice = samples[idx : idx + channels]
        if not channel_slice:
            continue
        mono_samples.append(int(sum(channel_slice) / len(channel_slice)))
    return encode_samples(mono_samples, sample_width)


def rms_chunk(chunk: bytes, sample_width: int) -> float:
    samples = decode_samples(chunk, sample_width)
    if not samples:
        return 0.0
    power = sum(float(sample) * float(sample) for sample in samples) / len(samples)
    return math.sqrt(power)


def decode_samples(raw_pcm: bytes, sample_width: int) -> list[int]:
    if sample_width == 1:
        return [value - 128 for value in raw_pcm]
    if sample_width == 2:
        count = len(raw_pcm) // 2
        return list(struct.unpack("<" + "h" * count, raw_pcm[: count * 2]))
    if sample_width == 3:
        samples: list[int] = []
        for idx in range(0, len(raw_pcm) - 2, 3):
            chunk = raw_pcm[idx : idx + 3]
            value = int.from_bytes(chunk + (b"\xff" if chunk[2] & 0x80 else b"\x00"), "little", signed=True)
            samples.append(value)
        return samples
    if sample_width == 4:
        count = len(raw_pcm) // 4
        return list(struct.unpack("<" + "i" * count, raw_pcm[: count * 4]))
    raise ValueError(f"Unsupported sample width: {sample_width}")


def encode_samples(samples: list[int], sample_width: int) -> bytes:
    if sample_width == 1:
        return bytes(max(0, min(255, sample + 128)) for sample in samples)
    if sample_width == 2:
        return struct.pack("<" + "h" * len(samples), *samples)
    if sample_width == 3:
        raw = bytearray()
        for sample in samples:
            sample = max(-(1 << 23), min((1 << 23) - 1, sample))
            raw.extend(int(sample).to_bytes(3, "little", signed=True))
        return bytes(raw)
    if sample_width == 4:
        return struct.pack("<" + "i" * len(samples), *samples)
    raise ValueError(f"Unsupported sample width: {sample_width}")

File: src/core/silence.ts
import { createRequire } from "node:module";
import { readFileSync } from "node:fs";
import type { SilenceSettings } from "./config.js";
import { timeRangeSeconds, type TimeRangeSeconds, timeRangeDuration } from "./models.js";

const require = createRequire(import.meta.url);
const { decode } = require("node-wav") as {
  decode: (buffer: Buffer) => { sampleRate: number; channelData: Float32Array[] };
};

/** Decode WAV to mono int16 PCM (matches Python wave + to_mono for silence analysis). */
export function readWavPcm(path: string): { sampleRate: number; sampleWidth: number; raw: Buffer } {
  const buf = readFileSync(path);
  const decoded = decode(buf);
  const channels = decoded.channelData.length;
  const n = decoded.channelData[0]?.length ?? 0;
  const raw = Buffer.allocUnsafe(n * 2);
  for (let i = 0; i < n; i++) {
    let s = 0;
    for (let c = 0; c < channels; c++) {
      s += decoded.channelData[c][i]!;
    }
    s /= channels;
    const int16 = Math.max(-32768, Math.min(32767, Math.round(s * 32767)));
    raw.writeInt16LE(int16, i * 2);
  }
  return { sampleRate: decoded.sampleRate, sampleWidth: 2, raw };
}

export function rmsToDb(rmsValue: number): number {
  if (rmsValue <= 0) return -120.0;
  const max16Bit = 32768.0;
  const normalized = Math.min(1.0, rmsValue / max16Bit);
  return 20.0 * Math.log10(Math.max(normalized, 1e-9));
}

export function detectSilenceRanges(
  sampleRate: number,
  sampleWidth: number,
  rawPcm: Buffer,
  settings: SilenceSettings,
): TimeRangeSeconds[] {
  const frameSamples = Math.max(1, Math.floor((sampleRate * settings.analysisFrameMs) / 1000.0));
  const frameBytes = frameSamples * sampleWidth;
  const minSilenceSeconds = settings.minSilenceDurationMs / 1000.0;

  const frameLoudnessDb: number[] = [];
  for (let idx = 0; idx < rawPcm.length; idx += frameBytes) {
    const chunk = rawPcm.subarray(idx, idx + frameBytes);
    if (chunk.length === 0) continue;
    const rmsVal = rmsChunk(chunk, sampleWidth);
    frameLoudnessDb.push(rmsToDb(rmsVal));
  }

  const silentRanges: TimeRangeSeconds[] = [];
  let currentStart: number | null = null;
  const frameMs = settings.analysisFrameMs;

  for (let frameIndex = 0; frameIndex < frameLoudnessDb.length; frameIndex++) {
    const loudnessDb = frameLoudnessDb[frameIndex]!;
    const isSilent = loudnessDb <= settings.silenceThresholdDb;
    if (isSilent && currentStart === null) {
      currentStart = frameIndex;
    }
    if (!isSilent && currentStart !== null) {
      const startSec = (currentStart * frameMs) / 1000.0;
      const endSec = (frameIndex * frameMs) / 1000.0;
      if (endSec - startSec >= minSilenceSeconds) {
        silentRanges.push(timeRangeSeconds(startSec, endSec));
      }
      currentStart = null;
    }
  }

  if (currentStart !== null) {
    const startSec = (currentStart * frameMs) / 1000.0;
    const endSec = (frameLoudnessDb.length * frameMs) / 1000.0;
    if (endSec - startSec >= minSilenceSeconds) {
      silentRanges.push(timeRangeSeconds(startSec, endSec));
    }
  }

  return mergeRanges(silentRanges);
}

export function mergeRanges(ranges: TimeRangeSeconds[]): TimeRangeSeconds[] {
  if (ranges.length === 0) return [];
  const sorted = [...ranges].sort((a, b) => a.start - b.start);
  const merged: TimeRangeSeconds[] = [sorted[0]!];
  for (const rng of sorted.slice(1)) {
    const prev = merged[merged.length - 1]!;
    if (rng.start <= prev.end) {
      merged[merged.length - 1] = timeRangeSeconds(prev.start, Math.max(prev.end, rng.end));
    } else {
      merged.push(rng);
    }
  }
  return merged;
}

export function applyAttackReleaseToSilenceRanges(
  ranges: TimeRangeSeconds[],
  clipDurationSeconds: number,
  preAttackMs: number,
  postReleaseMs: number,
): TimeRangeSeconds[] {
  const preAttack = Math.max(0.0, preAttackMs / 1000.0);
  const postRelease = Math.max(0.0, postReleaseMs / 1000.0);
  const adjusted = ranges
    .map((rng) =>
      timeRangeSeconds(
        Math.max(0.0, rng.start + postRelease),
        Math.min(clipDurationSeconds, rng.end - preAttack),
      ),
    )
    .filter((rng) => timeRangeDuration(rng) > 0.0);
  return mergeRanges(adjusted);
}

/**
 * @deprecated Use applyAttackReleaseToSilenceRanges. Kept for compatibility.
 */
export function expandWithPadding(
  ranges: TimeRangeSeconds[],
  clipDurationSeconds: number,
  preAttackMs: number,
  postReleaseMs: number,
): TimeRangeSeconds[] {
  return applyAttackReleaseToSilenceRanges(
    ranges,
    clipDurationSeconds,
    preAttackMs,
    postReleaseMs,
  );
}

export function invertRanges(removeRanges: TimeRangeSeconds[], totalDurationSeconds: number): TimeRangeSeconds[] {
  if (totalDurationSeconds <= 0) return [];
  if (removeRanges.length === 0) {
    return [timeRangeSeconds(0.0, totalDurationSeconds)];
  }

  const keeps: TimeRangeSeconds[] = [];
  let cursor = 0.0;
  for (const rng of removeRanges) {
    if (rng.start > cursor) {
      keeps.push(timeRangeSeconds(cursor, Math.min(rng.start, totalDurationSeconds)));
    }
    cursor = Math.max(cursor, rng.end);
  }

  if (cursor < totalDurationSeconds) {
    keeps.push(timeRangeSeconds(cursor, totalDurationSeconds));
  }

  return keeps.filter((k) => timeRangeDuration(k) > 0.0);
}

export function pruneShortKeepRanges(
  keepRanges: TimeRangeSeconds[],
  minKeepDurationMs: number,
): TimeRangeSeconds[] {
  const minKeepSeconds = Math.max(0, minKeepDurationMs) / 1000.0;
  if (minKeepSeconds <= 0 || keepRanges.length === 0) {
    return keepRanges;
  }

  const filtered = keepRanges.filter((rng) => timeRangeDuration(rng) >= minKeepSeconds);
  if (filtered.length > 0) {
    return filtered;
  }

  // Safety valve: avoid deleting a clip entirely when all keeps are short.
  let longest = keepRanges[0]!;
  for (const rng of keepRanges.slice(1)) {
    if (timeRangeDuration(rng) > timeRangeDuration(longest)) {
      longest = rng;
    }
  }
  return [longest];
}

function sampleCount(rawPcm: Buffer, sampleWidth: number): number {
  if (sampleWidth <= 0) return 0;
  return Math.floor(rawPcm.length / sampleWidth);
}

function readSampleAt(rawPcm: Buffer, sampleWidth: number, index: number): number {
  if (sampleWidth === 1) {
    return rawPcm[index]! - 128;
  }
  if (sampleWidth === 2) {
    return rawPcm.readInt16LE(index * 2);
  }
  if (sampleWidth === 3) {
    const offset = index * 3;
    let v = rawPcm[offset]! | (rawPcm[offset + 1]! << 8) | (rawPcm[offset + 2]! << 16);
    if (v & 0x800000) v |= ~0xffffff;
    return v;
  }
  if (sampleWidth === 4) {
    return rawPcm.readInt32LE(index * 4);
  }
  throw new Error(`Unsupported sample width: ${sampleWidth}`);
}

function hasZeroCrossing(rawPcm: Buffer, sampleWidth: number, index: number): boolean {
  if (index <= 0) return false;
  const prev = readSampleAt(rawPcm, sampleWidth, index - 1);
  const curr = readSampleAt(rawPcm, sampleWidth, index);
  if (prev === 0 || curr === 0) return true;
  return (prev < 0 && curr > 0) || (prev > 0 && curr < 0);
}

function nearestZeroCrossingIndex(
  rawPcm: Buffer,
  sampleWidth: number,
  targetIndex: number,
  searchSamples: number,
  totalSamples: number,
): number {
  if (totalSamples < 2 || targetIndex <= 0 || targetIndex >= totalSamples) {
    return targetIndex;
  }
  const clampedTarget = Math.max(1, Math.min(totalSamples - 1, targetIndex));
  const lo = Math.max(1, clampedTarget - searchSamples);
  const hi = Math.min(totalSamples - 1, clampedTarget + searchSamples);

  let bestCrossingIndex = -1;
  let bestCrossingDistance = Number.POSITIVE_INFINITY;
  let bestNearZeroIndex = clampedTarget;
  let bestNearZeroAmplitude = Number.POSITIVE_INFINITY;

  for (let idx = lo; idx <= hi; idx++) {
    const distance = Math.abs(idx - clampedTarget);
    const amplitude = Math.abs(readSampleAt(rawPcm, sampleWidth, idx));
    if (
      amplitude < bestNearZeroAmplitude ||
      (amplitude === bestNearZeroAmplitude &&
        distance < Math.abs(bestNearZeroIndex - clampedTarget))
    ) {
      bestNearZeroAmplitude = amplitude;
      bestNearZeroIndex = idx;
    }
    if (hasZeroCrossing(rawPcm, sampleWidth, idx) && distance < bestCrossingDistance) {
      bestCrossingDistance = distance;
      bestCrossingIndex = idx;
    }
  }

  return bestCrossingIndex >= 0 ? bestCrossingIndex : bestNearZeroIndex;
}

export function snapRangesToZeroCrossings(
  keepRanges: TimeRangeSeconds[],
  rawPcm: Buffer,
  sampleRate: number,
  sampleWidth: number,
  zeroCrossingSearchMs: number,
): TimeRangeSeconds[] {
  const totalSamples = sampleCount(rawPcm, sampleWidth);
  if (
    keepRanges.length === 0 ||
    sampleRate <= 0 ||
    totalSamples < 2 ||
    zeroCrossingSearchMs <= 0
  ) {
    return keepRanges;
  }
  const searchSamples = Math.max(
    1,
    Math.floor((Math.max(0, zeroCrossingSearchMs) * sampleRate) / 1000.0),
  );
  const clipDurationSeconds = totalSamples / sampleRate;

  const adjusted = keepRanges
    .map((rng) => {
      const startIdx = Math.max(0, Math.min(totalSamples, Math.round(rng.start * sampleRate)));
      const endIdx = Math.max(0, Math.min(totalSamples, Math.round(rng.end * sampleRate)));
      if (endIdx <= startIdx) {
        return null;
      }
      const snappedStartIdx =
        startIdx === 0
          ? 0
          : nearestZeroCrossingIndex(
              rawPcm,
              sampleWidth,
              startIdx,
              searchSamples,
              totalSamples,
            );
      const snappedEndIdx =
        endIdx === totalSamples
          ? totalSamples
          : nearestZeroCrossingIndex(
              rawPcm,
              sampleWidth,
              endIdx,
              searchSamples,
              totalSamples,
            );
      const adjustedStart = Math.max(0, Math.min(clipDurationSeconds, snappedStartIdx / sampleRate));
      const adjustedEnd = Math.max(0, Math.min(clipDurationSeconds, snappedEndIdx / sampleRate));
      if (adjustedEnd <= adjustedStart) {
        return null;
      }
      return timeRangeSeconds(adjustedStart, adjustedEnd);
    })
    .filter((rng): rng is TimeRangeSeconds => rng !== null);

  return mergeRanges(adjusted).filter((rng) => timeRangeDuration(rng) > 0.0);
}

export function rmsChunk(chunk: Buffer, sampleWidth: number): number {
  const samples = decodeSamples(chunk, sampleWidth);
  if (samples.length === 0) return 0.0;
  let power = 0;
  for (const sample of samples) {
    power += sample * sample;
  }
  power /= samples.length;
  return Math.sqrt(power);
}

export function decodeSamples(rawPcm: Buffer, sampleWidth: number): number[] {
  if (sampleWidth === 1) {
    return [...rawPcm].map((value) => value - 128);
  }
  if (sampleWidth === 2) {
    const count = Math.floor(rawPcm.length / 2);
    const out: number[] = [];
    for (let i = 0; i < count; i++) {
      out.push(rawPcm.readInt16LE(i * 2));
    }
    return out;
  }
  if (sampleWidth === 3) {
    const samples: number[] = [];
    for (let idx = 0; idx <= rawPcm.length - 3; idx += 3) {
      const b0 = rawPcm[idx]!;
      const b1 = rawPcm[idx + 1]!;
      const b2 = rawPcm[idx + 2]!;
      let v = b0 | (b1 << 8) | (b2 << 16);
      if (v & 0x800000) v |= ~0xffffff;
      samples.push(v);
    }
    return samples;
  }
  if (sampleWidth === 4) {
    const count = Math.floor(rawPcm.length / 4);
    const out: number[] = [];
    for (let i = 0; i < count; i++) {
      out.push(rawPcm.readInt32LE(i * 4));
    }
    return out;
  }
  throw new Error(`Unsupported sample width: ${sampleWidth}`);
}

export function encodeSamples(samples: number[], sampleWidth: number): Buffer {
  if (sampleWidth === 1) {
    return Buffer.from(samples.map((sample) => Math.max(0, Math.min(255, sample + 128))));
  }
  if (sampleWidth === 2) {
    const buf = Buffer.allocUnsafe(samples.length * 2);
    for (let i = 0; i < samples.length; i++) {
      buf.writeInt16LE(samples[i]!, i * 2);
    }
    return buf;
  }
  if (sampleWidth === 3) {
    const buf = Buffer.allocUnsafe(samples.length * 3);
    let o = 0;
    for (let sample of samples) {
      sample = Math.max(-(1 << 23), Math.min((1 << 23) - 1, sample));
      buf.writeIntLE(sample, o, 3);
      o += 3;
    }
    return buf;
  }
  if (sampleWidth === 4) {
    const buf = Buffer.allocUnsafe(samples.length * 4);
    for (let i = 0; i < samples.length; i++) {
      buf.writeInt32LE(samples[i]!, i * 4);
    }
    return buf;
  }
  throw new Error(`Unsupported sample width: ${sampleWidth}`);
}
